gpfit/constraint_set.py

```diff
@@ -0,0 +1,142 @@
+        elif fitdata["ftype"] == "SMA":
+            # constraint of the form w^alpha >= c1*u1^exp1 + c2*u2^exp2 +....
+            alpha = fitdata["a1"]
+            lhs, rhs = (ivar / self.mfac) ** alpha, NomialArray(monos).sum(0)
+        elif fitdata["ftype"] == "MA":
+            # constraint of the form w >= c1*u1^exp1, w >= c2*u2^exp2, ....
+            lhs, rhs = (ivar / self.mfac), NomialArray(monos).T
+
+        if fitdata["K"] == 1:
+            # when possible, return an equality constraint
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh == rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = lhs == rhs
+            else:
+                self.constraint = lhs == rhs
+        else:
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh >= rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = lhs >= rhs
+            else:
+                self.constraint = lhs >= rhs
+
+        self.bounds = {}
+        for i, dvar in enumerate(self.dvars):
+            self.bounds[dvar] = [fitdata["lb%d" % i], fitdata["ub%d" % i]]
+
+        ConstraintSet.__init__(self, [self.constraint])
+
+    def get_fitdata(self):
+        "return fit data"
+        return self.fitdata
+
+    def get_dataframe(self):
+        "return a pandas DataFrame of fit parameters"
+        import pandas as pd
+
+        df = pd.DataFrame(list(self.fitdata.values())).transpose()
+        df.columns = list(self.fitdata.keys())
+        return df
+
+    def process_result(self, result):
+        """
+        make sure fit result is within bounds of fitted data
+        """
+        super(FitConstraintSet, self).process_result(result)
+
+        if self.mfac not in result["sensitivities"]["constants"]:
+            return
+        if amax([abs(result["sensitivities"]["constants"][self.mfac])]) < 1e-5:
+            return
+
+        for dvar in self.dvars:
+            if isinstance(dvar, NomialArray):
+                num = [result(x) for x in dvar]
+            else:
+                num = result(dvar)
+            direct = None
+            if any(x < self.bounds[dvar][0] for x in hstack([num])):
+                direct, state = "lower", "below"
+                bnd = self.bounds[dvar][0]
+            if any(x > self.bounds[dvar][1] for x in hstack([num])):
+                direct, state = "upper", "above"
+                bnd = self.bounds[dvar][1]
+
+            if direct:
+                msg = (
+                    "Variable %.100s could cause inaccurate result"
+                    " because it is %s" % (dvar, state)
+                    + " %s bound. Solution is %.4f but"
+                    " bound is %.4f" % (direct, amax([num]), bnd)
+                )
+                print("Warning: " + msg)
```

Comment:
Is this what you're envisioning @bqpd?
```python
            if direct:                                                            
                msg = (                                                           
                    "Variable %.100s could cause inaccurate result"               
                    " because it is %s" % (dvar, state)                           
                    + " %s bound. Solution is %.4f but"                           
                    " bound is %.4f" % (direct, amax([num]), bnd)                 
                )                                                                 
                appendsolwarning(msg, self, result, "Fit Out-of-Bounds")
```
Is data (the second arg) allowed to be a ConstraintSet? In Tight it's a single constraint..

Reply:
it can be anything! it's discarded on save so it doesn't have to be pickleable. This looks good to me...double check it shows up in solution tables?

Reply:
Hmmm
```python
Traceback (most recent call last):
  File "/Users/philippe.kirschen/Optimization/gpfit/gpfit/tests/t_constraint_set.py", line 29, in test_fit_constraint_set_outside_bounds
    sol = m.solve()
  File "/Users/philippe.kirschen/Optimization/gpkit/gpkit/constraints/prog_factories.py", line 131, in solvefn
    self.process_result(result)
  File "/Users/philippe.kirschen/Optimization/gpkit/gpkit/constraints/set.py", line 188, in process_result
    constraint.process_result(result)
  File "/Users/philippe.kirschen/Optimization/gpfit/gpfit/constraint_set.py", line 138, in process_result
    appendsolwarning(msg, self, result, "Fit Out-of-Bounds")
  File "/Users/philippe.kirschen/Optimization/gpkit/gpkit/small_scripts.py", line 29, in appendsolwarning
    result["warnings"][category].append((msg, data))
KeyError: 'warnings'
```

Reply:
Ah duh. Yup prints! Added a unit test.